README.md

```diff
@@ -1,18 +1,84 @@
+to provide the best of both worlds.(If you are not familiar with using dagger, please take sometime to learn it, because it is a very good tool to reduce complexity, testing, and other software development merits)
+
+### Different use cases:
+1. If you just simply want to play a video without any interruption, you can just use ***DoubleViewTubiPlayerActivity*** out of box, then the code start a self contained activity to handle your playback experience
+```java
+      String subs = "http://put_your_own_subtitle.srt";
+                 String artwork = "http://www.put_your_own_art_work.png";
+                 String name = "Example Video";
+                 String video_url = "http://put_your_own_video_url.mp4";
+                 Intent intent = new Intent(SelectionActivity.this, DoubleViewTubiPlayerActivity.class);
+                 intent.putExtra(TubiPlayerActivity.TUBI_MEDIA_KEY, MediaModel.video(name, video_url, artwork, null));
+                 startActivity(intent);
+```
+
+2. The playback is still handle by standalone **DoubleViewTubiPlayerActivity** by using the above usage. In addition to play a video, if you also want to implement pre-roll, and middle-roll video ads, then there are more steps needed:
+    1. First the presumption is all the video ads need to be real time and dynamic, meaning all the ads are not fetched until they are needed.
+    2. Therefore, You need to know at beginning of playback at which positions has opportunity to show ads, in another words the list of positions during the main video to fetch ads from server, we called it the ***CuePoints***.
+    3. Then, when the main video proceed to cuepoints, you need to fetch the ads from your desired repository, your repository can return any number of ads to display during this cuepoint. (0 - any)
+    4. The **AdInterface** has two callbacks, since those there are ***RetrieveAdCallback*** callback in method params, you can choose to implement your own **asynchronous** implementation. You can use **AdRetriever** and **CuePointsRetriever** for customized fetching functionality. 
+        1. ***void fetchAd(AdRetriever retriever, RetrieveAdCallback callback)*** 
+        2. ***void fetchQuePoint(CuePointsRetriever retriever, CuePointCallBack callBack)***
+        
+
+
+
+## Player Lifecycle
+The ExoPlayer needs to create and release resources that binds to Activity's lifecycle, and also you may need to do some customized code logic in player's lifecycle.
+![State Machine](./documentation/tubiplayer_lifecycle.png)
+
+## State Machine Diagram
+The business logic of switching ads are complicated and stateful, in order to provide better state management, this diagrams below demonstrate the state transitions and you can use 
+this diagram to add your customized logic
```

Comment:
The business logic of switching ads is complicated and stateful.  In order to provide better state management, the diagrams below demonstrate the state transitions.  Refer to this
this diagram when adding your customized logic: